electrumsv/network_support/esv_client.py

```diff
@@ -0,0 +1,436 @@
+                    else:
+                        logger.error(f"PeerChannelMessage not recognised: {pc_message}")
+
+                if pc_message['content_type'] == 'application/octet-stream':
+                    logger.error(f"Binary format PeerChannelMessage received - not supported yet")
+        else:
+            logger.error("No messages could be returned from channel_id: %s, "
+                         "do you have a valid read token?", channel_id)
+
+    async def _message_fetcher_job(self) -> None:
+        """Idempotent - if spawned twice, the second time will do nothing"""
+        if not self._message_fetcher_is_alive:
+            peer_channel_notification_queue: asyncio.Queue[GeneralNotification] = asyncio.Queue()
+            for i in range(self._FETCH_JOBS_COUNT):
+                asyncio.create_task(
+                    self._fetch_peer_channel_message_job(peer_channel_notification_queue))
+
+            async for notification in self.subscribe_to_general_notifications():
+                peer_channel_notification_queue.put_nowait(notification)
+
+    async def wait_for_merkle_proofs_and_double_spends(self) -> AsyncIterable[TSCMerkleProof]:
+        if not self._message_fetcher_is_alive:
+            asyncio.create_task(self._message_fetcher_job())
+
+        # https://github.com/bitcoin-sv-specs/brfc-merchantapi#callback-notifications
+        while True:
+            # Todo run select query on MAPIBroadcastCallbacks to get libsodium encryption key
+            callback_response = await self._merkle_proofs_queue.get()
+            tsc_merkle_proof: TSCMerkleProof = json.loads(callback_response['callbackPayload'])
+
+            # Todo caller to delete entry in MAPIBroadcastCallbacks when processed
+            yield tsc_merkle_proof
+
+    async def subscribe_to_general_notifications(self) -> AsyncIterable[GeneralNotification]:
+        """Concurrent fetching of peer channel messages is left to the caller in order to keep
+        this class very simple"""
+        ws_base_url = self._replace_http_with_ws(self.base_url)
+        url = ws_base_url + "api/v1/web-socket" + f"?token={self.master_token}"
+        logger.debug(f"URL IS: {url}")
+        try:
+            async with self.session.ws_connect(url, headers={}, timeout=5.0) as ws:
+                logger.info(f'Connected to {url}')
+                msg: aiohttp.WSMessage
+                async for msg in ws:
+                    if msg.type == aiohttp.WSMsgType.TEXT:
+                        notification: GeneralNotification = json.loads(msg.data)
+                        yield notification
+
+                    if msg.type in (aiohttp.WSMsgType.CLOSE, aiohttp.WSMsgType.ERROR,
+                            aiohttp.WSMsgType.CLOSED, aiohttp.WSMsgType.CLOSING):
+                        logger.info("CLOSED")
+                        break
+        except WSServerHandshakeError as e:
+            if e.status == 401:
```

Comment:
Why is this ignoring other statuses? Comment?

Reply:
My bad. Fixed